x/smart/stream_dialer.go

```diff
@@ -460,7 +470,27 @@ func (f *StrategyFinder) parseConfig(configBytes []byte) (configConfig, error) {
 // It returns an error if no strategy was found that unblocks the testDomains.
 // The testDomains must be domains with a TLS service running on port 443.
 func (f *StrategyFinder) NewDialer(ctx context.Context, testDomains []string, configBytes []byte) (transport.StreamDialer, error) {
-	var parsedConfig configConfig
+	if winner, ok := unmarshalWinningStrategyFromCache(f.Cache); ok {
```

Comment:
We should only use the cached strategy if it's still in the input config.
If the config has changed, and the strategy is no longer there, we shouldn't use it anymore.

For DNS and TLS, you can simply move the saved ones to the top (if present), and proceed as usual.

For the fallback, you'd check if it exists in the fallback list, and test it before you do the DNS+TLS search.

Reply:
Done.